lessons/advanced/gen-stage.md

```diff
@@ -0,0 +1,261 @@
+
+As we've read, the role we give our stage is important.  The GenStage specification recognizes three roles:
+
++ `:producer` — A source.  Producers wait for demand from consumers and respond with the requested events.
+
++ `:producer_consumer` — Both a source and a sink.  Producer-consumers can respond to demand from other consumers as well as request events from producers.
+
++ `:consumer` — A sink.  A consumer requests and receives data from producers.
+
+Notice that our producers _wait_ for demand?  With GenStage our consumers send demand upstream and process the data from our producer.  This faciliates a mechanism known as back-pressure.  Back-pressure puts the onerous on the producer to not over-pressure when consumers are busy.
+
+Now that we've covered the roles within GenStage let's start on our application.
+
+## Getting Started
+
+In this example we'll be constructing a GenStage application that omits numbers, sort out the prime numbers, and finally print them.
+
+For our application we'll use all three GenStage roles.  Our producer will be responsible for counting and emits numbers.  We'll use a producer-consumer to filter out only the even numbers and later respond to demand from downsteam.  Last we'll build a consumer the remaining numbers for us.
+
+We'll begin by generating a project with a supervision tree:
+
+```shell
+$ mix new genstage_example --sup
+$ cd genstage_example
+```
+
+Let's update our dependencies in `mix.exs` to include `gen_stage`:
+
+```elixir
+  defp deps do
+    [
+      {:gen_stage, "~> 0.7},
+    ]
+  end
+```
+
+We should fetch our dependencies and compile before go much further:
+
+```shell
+$ mix do deps.get, compile
+```
+
+Now we're ready to build our producer!
+
+## Producer
+
+The first step of our GenStage application is creating our producer.  As we discussed before we want to create a producer that emits a constant stream of numbers.  Let's create our producer file:
+
+```shell
+$ mkdir lib/genstage_example
+$ touch lib/genstage_example/producer.ex
+```
+
+Now we can add the code:
+
+```
+defmodule GenstageExample.Producer do
+  alias Experimental.GenStage
+  
+  use GenStage
+  
+  def start_link(initial \\ 0) do
+    GenStage.start_link(__MODULE__, initial, name: __MODULE__)
+  end
+
+  def init(counter), do: {:producer, counter}
+
+  def handle_demand(demand, state) do
+    events = Enum.to_list(state..state + demand - 1)
+    {:noreply, events, (state + demand)}
+  end
+end
+```
+
+The two most important parts to take note of here are `init/1` and `handle_demand/2`.  In `init/1` we set the initial state as we've done in our GenServers but more importantly we label ourselves as a producer.  The response from our `init/1` function is what GenStage relies upon to classify our process.
+
+The `handle_demand/2` function is the real meat of our producer and must be implemented by all GenStage producers.  Here we return the set of numbers demanded by our consumers and increment our counter.
```

Comment:
`Meat` as used here might be difficult/weird to translate. (Thats on me, I wrote that line IIRC)